tests/lib/rules/no-octal-escape.js

```diff
@@ -58,9 +58,7 @@ ruleTester.run("no-octal-escape", rule, {
     ],
     invalid: [
 
-        // Test full message
-        { code: "var foo = \"foo \\01 bar\";", errors: [{ message: "Don't use octal: '\\01'. Use '\\u....' instead.", type: "Literal" }] },
```

Comment:
Looks like this might be an intentional test, but I don't think we do this for other rules so I'm not sure it's absolutely necessary. Curious what other teammates think.

Reply:
I think it is an intentional test, but I believe that defeats one of the whole points of using `messageId`s: (and this PR) you're now having to re-work your test if you want to change the actual message. Outside of the `message`, this test isn't testing anything not tested by the other tests.

I don't have a strong opinion, so happy to revert if the team wants, but that's also why I made this change in a few tests, to try and see what opinions came out of the woodwork :)

Reply:
Yeah, makes sense to me!

Reply:
It was my idea to add just 1 test with the full message, but don't have a strong opinion on is it really necessary.

> you're now having to re-work your test if you want to change the actual message.

On the other hand, someone could accidentally change the message in the code and tests wouldn't catch that.

It isn't just the text in the message. If someone now completely removes the `{{sequence}}` placeholder from the meta-message, only this test will fail.

Reply:
It could be also an unnoticed typo from the start:

```js
meta: {
    // ...
    messages: {
        octalEscapeSequence: "Don't use octal: '\\{{sequenc}}'. Use '\\u....' instead."
    }
}
```

The rule still works and all tests are passing except for that one.

Reply:
I agree, but it's also circular: 

> On the other hand, someone could accidentally change the message in the code and tests wouldn't catch that.

On the other hand, that change could be something you wanted, and now you're having to change it in at least two places.

>It isn't just the text in the message. If someone now completely removes the {{sequence}} placeholder from the meta-message, only this test will fail.

But is that actually a problem? I argue the problem is that you've now got an unused data property.

> The rule still works and all tests are passing except for that one.

It shouldn't, because the message won't be hydrated to the correct value?

Unless you're saying that the property is typod everywhere, such as it is in `sort-keys`:

https://github.com/eslint/eslint/blob/a5b3c5fa4edc2312534af0d9f0911f68144f8baf/lib/rules/sort-keys.js#L176

While sure it's nice not to have typos, they're not visible to the end user, and they're only catchable after the fact: the test will only fail when you stop making the typo.

-----

The issue imo is that there are actually two questions here: are we using the *indented* message?, and are we using an *understandable* message?

The former is something we can very easily write a test for: assign messages an id, and ensure that when we give the linter x, it uses message id y (which is what `messageId` is).

The latter problem however is a lot harder to solve, and is best done through PRs currently; in which case a single source of truth tends to be the best way to ensure easy reviewing :)

Reply:
I don't have a strong opinion either way, though I do think it would be nice to be consistent among our rules. Figuring out a nice way to test the readability/correctness of rule error messages could be an interesting problem to solve :)

Reply:
> On the other hand, that change could be something you wanted, and now you're having to change it in at least two places.

That wouldn't be so unusual, changes in the code usually require changes in the tests.

>> The rule still works and all tests are passing except for that one.

> It shouldn't, because the message won't be hydrated to the correct value?

Unfortunately, it does work. Try to change `{{sequence}}` to `{{sequenc}}` in the rule's code and all other tests will pass, but the user would be getting messages like:

 `Don't use octal: '\{{sequenc}}'. Use '\u....' instead.`

Reply:
> Unfortunately, it does work

Ahh yes I see because it does still match when hydrating; Yeah that's a nasty one.

Personally I'd say that the ideal is for the `RuleTester` to actually check for that; Would the team be open to `RuleTester` being able to error on unhydrated (or partial hydrated) messages? (i.e "error if a message contains {{ placeholder }}").

While I think it should the default, it might make sense to have it as an option, at least to allow disabling if you wanted that in your work 🤔

If that's desirable, I'm happy to try and fit it in for v7

Reply:
> That wouldn't be so unusual, changes in the code usually require changes in the tests.

For sure, but typically those are coupled by the type of change: logical changes require test changes. Likewise grammatical changes require editorial-style review 🤷‍♂

Meanwhile, you've just spotted a typo and now have to change two files instead of one - which is even more annoying in the Github editor.

(to be clear I'm in the "I don't have a strong opinion either way so long as it's consistent" boat too, but I think this is a good & interesting conversation 🙂 )

Reply:
> > Unfortunately, it does work
> 
> Ahh yes I see because it does still match when hydrating; Yeah that's a nasty one.

Yes, in the end the raw meta-message matches itself.

Both in the rule and the tests data properties remain unused and the place holder remains unused.

> Personally I'd say that the ideal is for the `RuleTester` to actually check for that; Would the team be open to `RuleTester` being able to error on unhydrated (or partial hydrated) messages? (i.e "error if a message contains {{ placeholder }}").
> 
> While I think it should the default, it might make sense to have it as an option, at least to allow disabling if you wanted that in your work 
> 
> If that's desirable, I'm happy to try and fit it in for v7

I think that something like this makes a lot of sense :+1: , but would probably need a RFC first.

Reply:
> would probably need a RFC first.

I had a feeling you'd say that - @kaicataldo would you agree?

I don't mind writing RFCs, but they can be draining & require a quorum😬

Reply:
> I don't have a strong opinion, so happy to revert if the team wants, but that's also why I made this change in a few tests, to try and see what opinions came out of the woodwork :)

I think I understand this now - if we don't remove this test, then it would probably make sense to leave some tests with full messages (e.g., one per file or one per messageId) in test files that are now entirely changed to messageIds in this PR?

Reply:
Something like that yes; once I'd converted all the rules to use `messageId`, I noticed some tests that used `message` but had `messageId`s.

I decided to stick with changing a few like this, but didn't have the energy to tackle the few that used `message` for every single test (so something like 50+ unique messages).

> it would probably make sense to leave some tests with full messages (e.g., one per file or one per messageId)

Exactly; which is where one of my points comes in: is that worth it? Given the messages are pretty stable, and that further changes to them tend to stand out vs the energy required to go through all of the rules and ensure that every rule has at least one test for each of their IDs, with valid placeholder values and inputs...

Reply:
Since we haven't been requiring that all tests have at least one test with the rendered error message, I'm of the mind that we should land this as it is and create a new issue to discuss further improvements. I don't have a strong opinion either way right now (I don't think this has been much of an issue in the past, though I could be wrong), but starting from a consistent place makes it easier to maintain and lays a foundation for further improvements.

Reply:
This is at the moment the only test in that rule that confirms that the user will see the detected octal escape sequence instead of nothing or a raw broken placeholder.

The test may become unnecessary in the future when RuleTester gets improved, but it does have a purpose now, at least I believe so.

As we don't yet have an opinion on the question of testing messages, what would be a clear benefit of removing this test now?

Reply:
My reasoning was mostly for the sake of consistency, but I’d rather not hold this PR up over this discussion. I think it’s fine to leave it in for now and continue the discussion afterwards. Thoughts?

Reply:
I agree that this conversation can be resolved for now.